### PR TITLE
[somfytahoma] Migrated to v2 Tahoma API

### LINKED ISSUE #6499: [somfytahoma] Login errors since Tahoma Update
I have login problems since the latest update on my Tahoma box (end of November):

```
2019-12-03 20:48:22.878 [hingStatusInfoChangedEvent] - 'somfytahoma:bridge:b5e4b0f3' changed from OFFLINE (COMMUNICATION_ERROR) to OFFLINE (COMMUNICATION_ERROR): Cannot get login cookie
2019-12-03 20:48:22.889 [hingStatusInfoChangedEvent] - 'somfytahoma:bridge:b5e4b0f3' changed from OFFLINE (COMMUNICATION_ERROR): Cannot get login cookie to OFFLINE (COMMUNICATION_ERROR)
2019-12-03 20:48:23.308 [hingStatusInfoChangedEvent] - 'somfytahoma:bridge:b5e4b0f3' changed from OFFLINE (COMMUNICATION_ERROR) to OFFLINE (COMMUNICATION_ERROR): Received invalid data
2019-12-03 20:48:23.322 [hingStatusInfoChangedEvent] - 'somfytahoma:bridge:b5e4b0f3' changed from OFFLINE (COMMUNICATION_ERROR): Received invalid data to OFFLINE (COMMUNICATION_ERROR)
```

I get these messages all over the time. Has anyone else this problem?

### PR DESCRIPTION
**Pull request by octa22**
Quite big update but very important to catch the OH 2.5 release,  because the current version of the binding stopped working due to API change.

Changelog:
- **migrated to v2 API** (fixes #6499)
- added support for the AdjuastableSlatsRollerShutter (#5866)
- added smoke sensor test feature (#5844)
- proper detection of gateway type
- proper handling of gateway state
- many other updates and performance improvements

### CODE REVIEW

**Comment by hakan42** [comment]
From my point of view, this change is necessary (because of the upstream API change).

The diff is quite big, though, so in-depth review might take more time than what we have until RC1 .

@openhab/2-x-add-ons-maintainers , @kaikreuzer , could we let this PR "slip by" ? The resulting jar is being tested by the community and has got positive reports.

**Comment by hakan42** [comment]
LGTM

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java [hunk 1]
```diff
@@ -153,28 +156,83 @@ public synchronized void login() {
...
+                    setTooManyRequests();
+                    return;
+                }
+            }
             logger.debug("Cannot get login cookie!", e);
             updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Cannot get login cookie");
+            if (e instanceof InterruptedException) {
+                Thread.currentThread().interrupt();
+            }
         } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/handler/SomfyTahomaBridgeHandler.java]: you are right

File: bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/adjustableslatsrollershutter.xml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="somfytahoma" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="adjustableslatsrollershutter">
+		<supported-bridge-type-refs>
+			<bridge-type-ref id="bridge"/>
+		</supported-bridge-type-refs>
+		<label>Somfy Adjustable Slats Roller Shutter</label>
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/adjustableslatsrollershutter.xml]: added description

File: bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/channels.xml [hunk 3]
```diff
@@ -238,4 +238,22 @@
...
+
+	<channel-type id="long_check">
+		<item-type>Switch</item-type>
+		<label>Smoke Sensor Long Check Trigger</label>
+		<description>A channel for triggering the smoke sensor's long check</description>
+	</channel-type>
+
+	<channel-type id="rocker">
+		<item-type>Switch</item-type>
+		<label>Rocker position</label>
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/channels.xml]: changed case

File: bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/channels.xml [hunk 4]
```diff
@@ -238,4 +238,22 @@
 		<label>Dock Long Beep Test</label>
 		<description>A channel for testing the dock's long beeping</description>
 	</channel-type>
+
+	<channel-type id="short_check">
+		<item-type>Switch</item-type>
+		<label>Smoke Sensor Short Check Trigger</label>
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/channels.xml]: done

File: bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/smokesensor.xml [hunk 5]
```diff
@@ -10,6 +10,8 @@
 		<label>Somfy Smoke Sensor</label>
 		<channels>
 			<channel id="contact" typeId="contact"></channel>
+			<channel id="short_check" typeId="short_check"></channel>
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/smokesensor.xml]: added

Comment (Hilbrand) [inline bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/smokesensor.xml]: These are not switches as they only send a command. They are better served by combining the short and long channels in a new channel and implement it using the command pattern, see: https://www.openhab.org/docs/developer/bindings/thing-xml.html#command-description

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/SomfyTahomaHandlerFactory.java [hunk 6]
```diff
@@ -49,6 +53,13 @@
 
     private Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegs = new HashMap<>();
 
+    private @NonNullByDefault({}) HttpClient httpClient;
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/SomfyTahomaHandlerFactory.java]: OK

File: bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/discovery/SomfyTahomaItemDiscoveryService.java [hunk 7]
```diff
@@ -239,6 +240,9 @@ private void discoverDevice(SomfyTahomaDevice device) {
             case SIREN:
                 deviceDiscovered(device, THING_TYPE_SIREN);
                 break;
+            case ADJUSTABLESLATSROLLERSHUTTER:
```
[the remark this answers is not in the document]
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/java/org/openhab/binding/somfytahoma/internal/discovery/SomfyTahomaItemDiscoveryService.java]: done

File: bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/smokesensor.xml (hunk 5 above)
Comment (octa22) [reply bundles/org.openhab.binding.somfytahoma/src/main/resources/ESH-INF/thing/smokesensor.xml]: changed. I hope I understood it corectly - cannot test it

**Comment by hoeckxer** [comment]
@octa22 any chance we can get this merged? Merge request has been approved, would be great to get into the master-branch for 2.5!

**Comment by hakan42** [comment]
Actually, @kaikreuzer , could we please get this into 2.5 ? The somfy devices are unsuable without this.

**Comment by kaikreuzer** [comment]
Absolutely - not sure why @Hilbrand didn't merge it right away when approving 👍 .

**Pull request merged**